File: Ninja Training/SpaceOptimization.py
def ninjaTraining(n, arr):
    dp = [ -1, -1 , -1]

    for i in range(3):
        dp[i] = arr[0][i]
    
    for i in range(1,n):
        dummy_dp = []
        for j in range(3):
            maximum = 0
            for k in range(3):
                if j != k:
                    maximum = max(maximum,arr[i][j] +  dp[k])
            dummy_dp.append(maximum)
        dp =  dummy_dp

    ans = 0
    for i in range(3):
        ans = max(ans,dp[i])
    return ans

if __name__ == "__main__":
    print(ninjaTraining(3,[ [18 ,11 ,19],[4, 13 ,7],[1, 8 ,13] ] ))

File: Ninja Training/SpaceOptimization.java



public class SpaceOptimization {
    public static void main(String[] args) {
        int[][] arr = {
            {18, 11, 19},
            {4, 13, 7},
            {1, 8, 13}
        };

        int size = arr.length;
        int[] dp = new int[3];

        System.arraycopy(arr[0],0,dp,0,3);

        for(int i = 1;i< size; i ++) {
            
            int[] dummy_arr = new int[3];
            for(int j = 0; j < 3; j++)  {
                int maximum = 0;

                for(int k = 0 ; k < 3; k ++) {
                    if( j != k) 
                        maximum = Math.max(maximum,arr[i][j] +  dp[k]);
                }

                dummy_arr[j] = maximum;
            }
            dp = dummy_arr;

        }

        int ans = 0;
        for(int i = 0; i< 3 ;i ++) {
            ans = Math.max(ans,dp[i]);
        }

        System.out.println(ans);
        
    }
}